src/serverenvironment.cpp

```diff
@@ -1070,6 +1070,74 @@ bool ServerEnvironment::swapNode(v3s16 p, const MapNode &n)
 	return true;
 }
 
+u8 ServerEnvironment::findSunlight(v3s16 pos)
+{
+	// Directions for neighbouring nodes with specified order
+	static const v3s16 dirs[] = {
+		v3s16(-1, 0, 0), v3s16(1, 0, 0), v3s16(0, 0, -1), v3s16(0, 0, 1),
+		v3s16(0, -1, 0), v3s16(0, 1, 0)
+	};
+
+	const NodeDefManager *ndef = m_server->ndef();
+	// found_light remembers the highest known sunlight value at pos
+	u8 found_light = 0;
+	struct stack_entry {
+		v3s16 pos;
+		s16 dist;
+	};
+	std::stack<stack_entry> stack;
+	stack.push({pos, 0});
+	std::unordered_map<u64, s8> dists;
+	dists[get_node_position_key(pos)] = 0;
+	while (!stack.empty()) {
+		struct stack_entry e = stack.top();
+		stack.pop();
+		v3s16 pos = e.pos;
+		s8 dist = e.dist + 1;
+		for (v3s16 off : dirs) {
+			v3s16 p = pos + off;
+			u64 h = get_node_position_key(p);
+			auto it = dists.find(h);
+			if (it != dists.end() && dist < it->second)
+				continue;
+
+			// Position to walk
+			bool is_position_ok;
+			MapNode node = getMap().getNodeNoEx(p, &is_position_ok);
+			if (!is_position_ok) {
+				// This happens very rarely because the map at pos is loaded
+				getMap().emergeBlock(p, false);
+				node = getMap().getNodeNoEx(p, &is_position_ok);
+				if (!is_position_ok)
+					continue;  // not generated
+			}
+			const ContentFeatures &def = ndef->get(node);
+			if (is_position_ok && def.sunlight_propagates) {
```

Comment:
I think, `is_position_ok` is never a false value here.

Reply:
fixed